includes/AMP/Traits/Sanitization_Utils.php

```diff
@@ -0,0 +1,182 @@
+ * @link      https://github.com/google/web-stories-wp
+ */
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories\AMP\Traits;
+
+use DOMElement;
+use Google\Web_Stories_Dependencies\AmpProject\Dom\Document;
+use \AmpProject\Dom\Document as AMP_Document;
+
+/**
+ * Trait Sanitization_Utils
+ *
+ * @since 1.1.0
+ */
+trait Sanitization_Utils {
+	/**
+	 * Replaces the HTML start tag to make the language attributes dynamic.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document Document instance.
+	 * @return void
+	 */
+	private function transform_html_start_tag( &$document ) {
+		$document->html->setAttribute( 'amp', '' );
+
+		// See get_language_attributes().
+		if ( is_rtl() ) {
+			$document->html->setAttribute( 'dir', 'rtl' );
+		}
+
+		$lang = get_bloginfo( 'language' );
+		if ( $lang ) {
+			$document->html->setAttribute( 'lang', esc_attr( $lang ) );
+		}
+	}
+
+	/**
+	 * Transform all a tags to add target and rel attributes.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document Document instance.
+	 * @return void
+	 */
+	private function transform_a_tags( &$document ) {
+		$hyperlinks = $document->getElementsByTagName( 'a' );
+		/* @var DOMElement $hyperlink The <a> element */
+		foreach ( $hyperlinks as $hyperlink ) {
+			if ( ! $hyperlink->getAttribute( 'target' ) ) {
+				$hyperlink->setAttribute( 'target', '_blank' );
+			}
+			if ( ! $hyperlink->getAttribute( 'rel' ) ) {
+				$hyperlink->setAttribute( 'rel', 'noreferrer' );
+			}
+		}
+	}
+
+	/**
+	 * Replaces the amp-story end tag to include amp-analytics tag if set up.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document Document instance.
+	 * @return void
+	 */
+	private function insert_analytics_configuration( &$document ) {
+		/* @var DOMElement $story_element The <amp-story> element. */
+		$story_element = $document->body->getElementsByTagName( 'amp-story' )->item( 0 );
+
+		if ( ! $story_element ) {
+			return;
+		}
+
+		ob_start();
+
+		/**
+		 * Fires before the closing <amp-story> tag.
+		 *
+		 * Can be used to print <amp-analytics> configuration.
+		 *
+		 * @since 1.1.0
+		 */
+		do_action( 'web_stories_print_analytics' );
+
+		$output = (string) ob_get_clean();
+
+		if ( empty( $output ) ) {
+			return;
+		}
+
+		$fragment = $document->createDocumentFragment();
+		$fragment->appendXml( $output );
```

Comment:
Is this safe? It requires that authors write XML rather than HTML here. For example, if someone tried to add `<amp-analytics type=gtag>` this would be an XML error, even though it is valid HTML. Should there be at least `@` suppression here? Otherwise, should this be parsed into a new HTML `Document` and then import the nodes to append to `$story_element`?

Reply:
Do you perhaps have an example of the latter?

Note that this code is already in the plugin code base today (in `HTML` class). We could solve this in a follow-up PR.

Reply:
I believe something like this will work:

```suggestion
		$fragment_document = Document::fromHtmlFragment( $output );
		while ( $fragment_document->body->firstChild ) {
			$node = $fragment_document->body->removeChild( $fragment_document->body->firstChild );
			$node = $document->importNode( $node, true );
			$fragment->appendChild( $node );
		}
```